e3db-crypto-android/src/main/java/com/tozny/e3db/AndroidCrypto.java

```diff
@@ -20,54 +20,82 @@
 
 package com.tozny.e3db;
 
+import com.goterl.lazycode.lazysodium.LazySodium;
+import com.goterl.lazycode.lazysodium.LazySodiumAndroid;
+import com.goterl.lazycode.lazysodium.exceptions.SodiumException;
+import com.goterl.lazycode.lazysodium.interfaces.Box;
+import com.goterl.lazycode.lazysodium.interfaces.SecretBox;
+import com.goterl.lazycode.lazysodium.interfaces.SecretStream;
+import com.goterl.lazycode.lazysodium.interfaces.Sign;
 import com.tozny.e3db.crypto.*;
-import org.libsodium.jni.NaCl;
-import org.libsodium.jni.Sodium;
-import org.libsodium.jni.crypto.Box;
-import org.libsodium.jni.crypto.Random;
-import org.libsodium.jni.crypto.SecretBox;
+
 import static com.tozny.e3db.Checks.*;
-import static org.libsodium.jni.Sodium.*;
+import com.goterl.lazycode.lazysodium.SodiumAndroid;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
 
 class AndroidCrypto implements Crypto {
-  private final Sodium sodium;
-  private final Random random;
+  private static final int BLOCK_SIZE = 65_536; // 2 ^ 16
 
+  private final LazySodium lazySodium;
+  private static final Charset UTF8 = Charset.forName("UTF-8");
   public AndroidCrypto() {
-    // Make sure libsodium initialization occurs.
-    sodium = SodiumInit.sodium;
-    random = new Random();
+    lazySodium = new LazySodiumAndroid(Init.sodium);
   }
 
-  private static class SodiumInit {
+  private static class Init {
+
     // Static inner class as a singleton to make sure
+    public static final SodiumAndroid sodium;
     // Sodium library is initalized once and only once.
-    public static Sodium sodium = NaCl.sodium();
+    static {
+      try {
+         sodium = new SodiumAndroid();
+      }
+      catch(Throwable ex) {
+        ex.printStackTrace();
+        throw new RuntimeException(ex);
+      }
+    }
   }
 
   @Override
   public CipherWithNonce encryptSecretBox(byte[] message, byte[] key) {
     checkNotNull(message, "message");
     checkNotEmpty(key, "key");
-    byte[] nonce = random.randomBytes(crypto_secretbox_noncebytes());
-    byte[] cipher = new SecretBox(key).encrypt(nonce, message);
+    byte[] nonce = lazySodium.randomBytesBuf(SecretBox.NONCEBYTES);
```

Comment:
Does lazySodium's PRNG rely on a secure PRNG or is it self-implemented?

Reply:
I think it relies on the underlying OS, but that is really a detail of libsodium rather than lazySodium.

Reply:
That works for me. Some of the other language bindings use Sodium's wrapper for randomness as a passthru to the actual PRNG on the OS. Other libraries, for whatever reason, try to use their own RNG that may or may not be cryptographically strong. Just wanted to make sure we were still using a safe RNG.